### PR TITLE
added runUntilFirstDone method to awex

### PR DESCRIPTION
**Pull request by alexmiragall**
Created new method untilFirstDone:

```
/**
+     * Creates a new promise that will be resolved with the result of the first task which completes correctly.
+     * Task will execute sequentially until the first one is resolved.
+     * If all the promises fail promise will be rejected with the last exception thrown.
+     * In case the collection of tasks is empty the promise will be rejected with an EmptyTasksException.
+     *
+     * @param tasks source tasks to execute sequentially
+     * @param <Result> type of result of the promises
+     * @param <Progress> type of progress of the promises
+     * @returna new promise that will be resolved with the first task completed correctly
+     */
```

Also added new pipeCallbacks to promises and the new PipePromise subclass. All these implementations are based on the jDeferred library.

### CODE REVIEW

**Comment by alexmiragall** [comment]
Tests are pending 👍

File: awex-core/src/main/java/com/raycoarana/awex/AwexPromise.java [hunk 1]
```diff
@@ -686,6 +689,14 @@ public void onCancel() {
         return promise;
     }
 
+    @Override
+    public <Result_Out, Progress_Out> Promise<Result_Out, Progress_Out> pipe(
```
Comment (raycoarana) [inline awex-core/src/main/java/com/raycoarana/awex/AwexPromise.java]: I don't like the idea of adding a new pipe() method, I think you could implement this using the current pipe() method, providing a Promise to it. You just then need to add a fail callback to that promise.

File: awex-core/src/main/java/com/raycoarana/awex/AwexPromise.java [hunk 2]
```diff
@@ -27,8 +30,8 @@
 class AwexPromise<Result, Progress> implements ResolvablePromise<Result, Progress> {
 
     protected final Awex mAwex;
+    protected final Task mTask;
```
Comment (raycoarana) [inline awex-core/src/main/java/com/raycoarana/awex/AwexPromise.java]: Why protected?

File: awex-core/src/main/java/com/raycoarana/awex/Awex.java [hunk 3]
```diff
@@ -114,6 +120,61 @@ void submit(Runnable runnable) {
...
+     * Task will execute sequentially until the first one is resolved.
+     * If all the promises fail promise will be rejected with the last exception thrown.
+     * In case the collection of tasks is empty the promise will be rejected with an EmptyTasksException.
+     *
+     * @param tasks source tasks to execute sequentially
+     * @param <Result> type of result of the promises
+     * @param <Progress> type of progress of the promises
+     * @returna new promise that will be resolved with the first task completed correctly
+     */
+    public <Result, Progress> Promise<Result, Progress> runUntilFirstDone(final
```
Comment (raycoarana) [inline awex-core/src/main/java/com/raycoarana/awex/Awex.java]: To not be confused with anyOf() operation, I would name it more explicitly about it's sequential. Maybe sequentiallyUntilFirstDone().

File: awex-core/src/main/java/com/raycoarana/awex/Awex.java [hunk 4]
```diff
@@ -1,9 +1,15 @@
...
+import com.raycoarana.awex.callbacks.DonePipeCallback;
+import com.raycoarana.awex.callbacks.FailCallback;
+import com.raycoarana.awex.callbacks.FailPipeCallback;
+import com.raycoarana.awex.callbacks.ProgressPipeCallback;
 import com.raycoarana.awex.exceptions.AbsentValueException;
+import com.raycoarana.awex.exceptions.EmptyTasksException;
 import com.raycoarana.awex.state.PoolStateImpl;
 import com.raycoarana.awex.state.QueueStateImpl;
 import com.raycoarana.awex.util.Map;
+import com.sun.javaws.exceptions.InvalidArgumentException;
```
Comment (raycoarana) [inline awex-core/src/main/java/com/raycoarana/awex/Awex.java]: You don't want that one :-)

File: awex-core/src/main/java/com/raycoarana/awex/AwexPromise.java (hunk 2 above)
Comment (alexmiragall) [inline awex-core/src/main/java/com/raycoarana/awex/AwexPromise.java]: fixed as I have removed the PipedPromise

File: awex-core/src/main/java/com/raycoarana/awex/Awex.java (hunk 3 above)
Comment (alexmiragall) [inline awex-core/src/main/java/com/raycoarana/awex/Awex.java]: changed :)

**Pull request merged**